test/e2e/flowcontrol/application_test.go

```diff
@@ -0,0 +1,158 @@
+			}
+		}
+
+	}
+
+	BeforeEach(func() {
+		logStressors = make([]corev1.Pod, 0)
+		e2e = framework.NewE2ETestFramework()
+
+		cl = helpers.NewClusterLogging(helpers.ComponentTypeCollectorVector)
+
+		lokiNS = e2e.CreateTestNamespace()
+		stressorNS = e2e.CreateTestNamespace()
+
+		e2e.AddCleanup(func() error {
+			opts := metav1.DeleteOptions{}
+			return e2e.KubeClient.CoreV1().Namespaces().Delete(context.TODO(), lokiNS, opts)
+		})
+
+		e2e.AddCleanup(func() error {
+			opts := metav1.DeleteOptions{}
+			return e2e.KubeClient.CoreV1().Namespaces().Delete(context.TODO(), stressorNS, opts)
+		})
+
+		l = loki.NewReceiver(lokiNS, "loki-server")
+
+		clf = runtime.NewClusterLogForwarder()
+		clf.Spec.Inputs = []loggingv1.InputSpec{
+			{
+				Name: "custom-app",
+				Application: &loggingv1.Application{
+					Namespaces: []string{stressorNS},
+				},
+			},
+		}
+		clf.Spec.Outputs = []loggingv1.OutputSpec{
+			{
+				Name: loggingv1.OutputTypeLoki,
+				Type: loggingv1.OutputTypeLoki,
+				URL:  l.InternalURL("").String(),
+				OutputTypeSpec: loggingv1.OutputTypeSpec{
+					Loki: &loggingv1.Loki{},
+				},
+			},
+		}
+		clf.Spec.Pipelines = []loggingv1.PipelineSpec{
+			{
+				InputRefs:  []string{"custom-app"},
+				OutputRefs: []string{loggingv1.OutputTypeLoki},
+				Name:       "flow-control",
+			},
+		}
+
+		for i := 0; i < 10; i += 1 {
+			logStressors = append(logStressors,
+				*runtime.NewLogGeneratorWithPodInf(stressorNS, fmt.Sprintf("stressor-%d", i), 10*time.Millisecond, "Test Message"),
+			) // 100 lines per second
+		}
+
+	})
+
+	AfterEach(func() {
+		e2e.Cleanup()
+	}, framework.DefaultCleanUpTimeout)
+
+	Describe("When source is application", func() {
+		Describe("using drop policy", func() {
+			It("setting group limit", func() {
+				clf.Spec.Inputs[0].Application.GroupLimit = &loggingv1.LimitSpec{
+					Policy:              loggingv1.DropPolicy,
+					MaxRecordsPerSecond: 100,
+				} // 10 files and 100 group limit, so 10 lines per file
+
+				DeployLoggingComponents()
+				if WaitForMetricsToShow() == false {
+					Fail("Metrics not showing up in Prometheus")
+				}
+				// sleeping for 2 minutes to ensure rate of metrics is stable
+				time.Sleep(2 * time.Minute)
+
+				promQuery = fmt.Sprintf(VectorCompSentEvents, FlowControlComponent)
+				ExpectMetricsWithinRange(GetCollectorMetrics(promQuery), 0, 101)
+			})
+
+			It("setting container limit", func() {
+				clf.Spec.Inputs[0].Application.ContainerLimit = &loggingv1.LimitSpec{
+					Policy:              loggingv1.DropPolicy,
+					MaxRecordsPerSecond: 100,
+				} // 10 files and 100 group limit, so 10 lines per file
+
+				DeployLoggingComponents()
+				if WaitForMetricsToShow() == false {
+					Fail("Metrics not showing up in Prometheus")
+				}
+				// sleeping for 2 minutes to ensure rate of metrics is stable
+				time.Sleep(2 * time.Minute)
+
+				promQuery = fmt.Sprintf(SumMetric, VectorCompSentEvents)
```

Comment:
This line has no effect.

Reply:
@alanconway Can you please explain why this line has no effect?

Reply:
I should have put the comment a line further down, so you can see the next line:
```
149				promQuery = fmt.Sprintf(SumMetric, VectorCompSentEvents)
150				promQuery = fmt.Sprintf(promQuery, FlowControlComponent)
```
Assigning twice to same variable so only second assignment has any effect.